README.md
# Pyroms

Welcome to Pyroms!

Pyroms is a collection of tools to process input and output files
from the Regional Ocean Modeling System, [ROMS](https://www.myroms.org/). It was originally
started by Rob Hetland as a googlecode project, then he morphed it
into octant, also at googlecode. Frederic Castruccio then created a
fork and renamed it back to pyroms.

Pyroms is now hosted on GitHub.com in the [ESMG/pyroms](https://github.com/ESMG/pyroms) repository. This version is on the [python3](https://github.com/ESMG/pyroms/tree/python3) branch. It requires Python 3.4 or later.

## Installation

Pyroms is still a bit rough around the edges, particularly with regard to installation. Recent development has been done in Python environments managed by [Conda](https://docs.conda.io/en/latest/). However Pyroms itself cannot yet be installed with Conda.

If you are starting from scratch, we recommend that you install
[Anaconda](https://www.anaconda.com/) or
[Miniconda](https://docs.conda.io/en/latest/miniconda.html) and create a Python 3 environment (as of December 2020, version 3.8 is your best bet) for Pyroms and your other scientific software. You should also consider making conda-forge your default channel. See the [conda-forge tips and tricks page](https://conda-forge.org/docs/user/tipsandtricks.html).

NB Joe McGovern - make conda-forge your default channel on conda

If you don't want to use Conda, that's fine, but you will have to do more of the work yourself.

## Prerequisites

The following are required and are all available from [Conda-Forge](https://conda-forge.org/).

   * Python >= 3.4 (Python 3.8 currently recommended for new environments)
   * [numpy](https://numpy.org/)
   * [scipy](https://www.scipy.org/)
   * [matplotlib](https://matplotlib.org/)
   * [basemap](https://matplotlib.org/basemap/)
   * [netcdf4](https://unidata.github.io/netcdf4-python/netCDF4/index.html)
   * [cftime](https://unidata.github.io/cftime/)
   * [lpsolve55]
     * conda install -c conda-forge lpsolve55
   * [pip](https://pypi.org/project/pip/)

If you experience an issue with the following dedent error (https://github.com/matplotlib/basemap/issues/494):
   * from matplotlib.cbook import dedent ==> "ImportError: cannot import name 'dedent'" in matplotlib
   * Open .../site-packages/mpl_toolkits/basemap/__init__.py at relevant breakpoint
   * replace from ... import statement with
     * from inspect import cleandoc as dedent
     * THIS WILL NEED TO BE REDONE AT EACH IDE LAUNCH DUE TO LACK OF OWNERSHIP CREDENTIALS FOR MATPLOTLIB/BASEMAP
   * Alternatively, execute: pip install -U matplotlib==3.2
   * If you get KeyError: 'PROJ_LIB' as part of basemap running/installation, see the following links
     * set os.environ['PROJ_LIB'] = '/home/pete/anaconda3/envs/pyroms_MI/share/proj'
       * or os.environ['PROJ_LIB'] = anaconda location + '/pyroms_MI/share/proj'
     * https://stackoverflow.com/questions/56313027/keyerror-proj-lib-while-installing-basemap-using-conda-install
     * https://stackoverflow.com/questions/59376657/proj-lib-error-in-spyder-with-importing-basemap
     * https://github.com/conda-forge/basemap-feedstock/issues/30#issuecomment-423512069

If you encounter a memory issue with meshgrid, try the following
replace
lon, lat = np.meshgrid(long, lat)
with
lon, lat = np.meshgrid(long, lat, sparse=True)

If you get OSError: no file with expected extension:
(related to pyroms.grid.Gridgen), in hgrid.py (compiler will probably open in debud mode)
replace
        self._libgridgen = np.ctypeslib.load_library('libgridgen.', pyroms.__path__[0]) 
with
        self._libgridgen = np.ctypeslib.load_library('libgridgen.so', pyroms.__path__[0]) 

uninstall and reinstall pyroms, pyroms_toolbox and bathy_smoother

If this doesn't work, try the following solution from Mike Hadfield, NZ
https://github.com/ESMG/pyroms/issues/26
1) conda install -c conda-forge gridgen
2) then add the following 
   a) np.ctypeslib.load_library('libgridgen','/home/pete/anaconda3/envs/pyroms_MI/lib')


The following is optional: Pyroms can be built and run without it but some of the functionality will be missing.

   * scrip, a Python implementation of [SCRIP](https://github.com/SCRIP-Project/SCRIP),
     the Spherical Coordinate Remapping and Interpolation Package. This is used by the pyroms
     module. The Python scrip code (a rather old version) is
     [bundled in pyroms](https://github.com/ESMG/pyroms/tree/python3/pyroms/external/scrip)
     and can be built and installed separately as described below. In future we plan to
     move from the bundled scrip code to a stand-alone package like
     [ESMF/ESMPy](https://www.earthsystemcog.org/projects/esmpy/) or
     [PySCRIP](https://github.com/dchandan/PySCRIP).

The following is optional and provides high-resolution coastlines for basemap:

   * [basemap-data-hires](https://anaconda.org/conda-forge/basemap-data-hires/)

## Install from source

To clone a copy of the source and install the pyroms packages, you can use the following commands
```
# Cd to a convenient directory
$ git clone https://github.com/ESMG/pyroms.git
$ pip install -e pyroms/pyroms
$ pip install -e pyroms/pyroms_toolbox
$ pip install -e pyroms/bathy_smoother
```

This installs three PIP packages with the names pyroms, pyroms\_toolbox and bathy\_smoother,
each with an [eponymous](https://en.wiktionary.org/wiki/eponymous) module.

An [editable-mode](https://pip.pypa.io/en/stable/reference/pip_install/#editable-installs) installation is recommended becauses it means changes you make to your copy of the source code will take effect when you import the modules. If you don't want this you can omit the "-e" option

The "pip install" command runs "python setup.py install" (or "python setup.py develop" with the "-e" switch) in each of the subdirectories listed. The "pip install" form is recommended because it allow easy removal (see below)

The above should work on most Linuces and on OSX with the system gcc and gfortran compilers.
They have also been verified to work in a Conda environment on Windows,
provided you install the
[m2w64-gcc](https://anaconda.org/msys2/m2w64-gcc) and [m2w64-gfortran](https://anaconda.org/msys2/m2w64-gcc-fortran) compilers.

## Install scrip

If you install as above and try to import the three Pyroms modules without having installed
scrip you will get a warning like this:

```
$ python
Python 3.8.5 | packaged by conda-forge | (default, Aug 29 2020, 01:22:49)
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyroms
WARNING:root: scrip could not be imported. Remapping functions will not be available
>>> import pyroms_toolbox
>>> import bathy_smoother
```

The scrip module is not available via Conda or any other package repository and we are looking at alternatives. In the meantime, scrip can be built and installed from source as follows

```
# Start in the directory into which you cloned pyroms and cd to the SCRIP
# source directory
$ cd pyroms/pyroms/external/scrip/source/

# Print the location of the active Conda environment (which is called "python38"
# in this case). The active environment location is used to find the netCDF and
# other libraries.
$ conda info | grep "active env location"
    active env location : /home/hadfield/miniconda3/envs/python38

# make sure that all requirements in the scrip makefile are installed
Consider this link, if issues with libhdf detection....
https://stackoverflow.com/questions/50593848/error-building-caffe-cannot-find-lhdf5-hl

also 

sudo apt-get install libhdf5-serial-dev netcdf-bin libnetcdf-dev

and this needed to be added to scrip makefile 

HDF_LIBDIR    ?= $(PREFIX)/lib/x86_64-linux-gnu/hdf5/serial

and change
LIB = -L$(SRCDIR) -L$(NETCDF_LIBDIR) $(NETCDF_LIBS)
to
LIB = -L$(SRCDIR) -L$(NETCDF_LIBDIR) -L$(HDF_LIBDIR) $(NETCDF_LIBS)

Needed to replace ?= with :=

Tried 
sudo apt-get install libtool

Needed to create a new symlink in /usr/lib using the following command
sudo ln -s /usr/lib/libdf.so.0.0.0 /usr/lib/libdf.so

# Run make to build the scrip Python extension and install it into the Conda
# environment. The makefile calculates a variable called SCRIP_EXT_DIR, into
# which it installs the scrip Python extension. If pyroms has been installed
# in editable (development) mode, set the DEVELOP variable to a non-empty value.
$ export PREFIX=/home/hadfield/miniconda3/envs/python38
$ make DEVELOP=1 PREFIX=$PREFIX install
$ mv -vf scrip*.so ../../../pyroms
‘scrip.cpython-38-x86_64-linux-gnu.so’ -> ‘../../../pyroms/scrip.cpython-38-x86_64-linux-gnu.so’
```

## Removal

To remove the three Pyroms packages you can use the "pip uninstall" command, referring to the packages by their package names

```
# Run from any directory in the same environment as you installed
# and use the package name
$ pip uninstall pyroms
$ pip uninstall pyroms_toolbox
$ pip uninstall bathy_smoother
```

If you have built and installed the scrip extension from the makefile as above, you can also uninstall it with the makefile. The PREFIX does not need to be set in this case.

```
# Start in the directory into which you cloned pyroms and cd to the SCRIP
# source directory
$ cd pyroms/pyroms/external/scrip/source/

# Remove with make.
$ make DEVELOP=1 uninstall
```

## Running Pyroms

We have a gridid.txt file that's pointed to by the PYROMS\_GRIDID\_FILE
environment variable. If you are operating on files containing
sufficient grid information already, you won't need to use this.
An example is provided in the examples directory.


## Doxygen

Running "doxygen .doxygen" in any of pyroms, pyroms\_toolbox or
bathy\_smoother will generate doxygen files. Edit the .doxygen files to
specify html vs. some other output format.


# Notes JVMCGOVERN
# To get lpsolve working, Try:
# 1: (didn't work)
#  conda install -c snorfalorpagus lpsolve
# 2:
# https://stackoverflow.com/questions/14906603/how-to-install-lpsolve-module-for-python-on-linux-ubuntu-10-04

CELTIC_SEA/history.txt
    1  2022-10-17 17:53:14 nano src/mod_akima_1d.f90
    2  2022-10-17 17:53:14 make
    3  2022-10-17 17:53:14 nano src/mod_akima_1d.f90
    4  2022-10-17 17:53:14 sudo apt-get install gfortran
    5  2022-10-17 17:53:14 sudo apt-get install gfortran-9
    6  2022-10-17 17:53:14 sudo apt update
    7  2022-10-17 17:53:14 sudo apt list --upgradable
    8  2022-10-17 17:53:14 sudo apt install gfortran-9
    9  2022-10-17 17:53:14 sudo apt-get update && sudo apt-get upgrade
   10  2022-10-17 17:53:14 sudo apt install gfortran-9
   11  2022-10-17 17:53:14 sudo add-apt-repository universe
   12  2022-10-17 17:53:14 sudo add-apt-repository "deb http://archive.ubuntu.com/ubuntu $(lsb_release -sc) main universe restricted multiverse"
   13  2022-10-17 17:53:14 sudo apt install gfortran-9
   14  2022-10-17 17:53:14 sudo apt install gfortran
   15  2022-10-17 17:53:14 sudo apt-get update
   16  2022-10-17 17:53:14 sudo apt install gfortran
   17  2022-10-17 17:53:14 sudo apt install gfortran-9
   18  2022-10-17 17:53:14 sudo apt install gcc
   19  2022-10-17 17:53:14 sudo add-apt-repository universe
   20  2022-10-17 17:53:14 sudo apt-get update
   21  2022-10-17 17:53:14 sudo add-apt-repository universe
   22  2022-10-17 17:53:14 sudo apt install gfortran-9
   23  2022-10-17 17:53:14 apt install gfortran-9
   24  2022-10-17 17:53:14 root
   25  2022-10-17 17:53:14 sudo add-apt-repository partner
   26  2022-10-17 17:53:14 sudo apt-get update && sudo apt-get upgrade
   27  2022-10-17 17:53:14 sudo apt install gfortran-9
   28  2022-10-17 17:53:14 sudo apt install gfortran-7
   29  2022-10-17 17:53:14 sudo apt install gfortran-6
   30  2022-10-17 17:53:14 sudo apt install gfortran-5
   31  2022-10-17 17:53:14 sudo apt-mark auto gfortran
   32  2022-10-17 17:53:14 sudo apt install gfortran-9
   33  2022-10-17 17:53:14 sudo apt install gfortran-6
   34  2022-10-17 17:53:14 sudo apt install gfortran-5
   35  2022-10-17 17:53:14 sudo apt update
   36  2022-10-17 17:53:14 sudo apt upgrade gfortran
   37  2022-10-17 17:53:14 sudo apt upgrade gfortran-9
   38  2022-10-17 17:53:14 sudo apt update gfortran
   39  2022-10-17 17:53:14 gfortran-9 --version
   40  2022-10-17 17:53:14 sudo add-apt-repository ppa:ubuntu-toolchain-r/test
   41  2022-10-17 17:53:14 sudo apt-get update
   42  2022-10-17 17:53:14 sudo apt install gfortran-9
   43  2022-10-17 17:53:14 gfortran --version
   44  2022-10-17 17:53:14 gfortran-9 --version
   45  2022-10-17 17:53:14 sudo update-alternatives
   46  2022-10-17 17:53:14 gfortran-9 --all
   47  2022-10-17 17:53:14 gfortran-9
   48  2022-10-17 17:53:14 sudo update-alternatives --install /usr/bin/gfortran gfortran /usr/bin/gfortran-9 60
   49  2022-10-17 17:53:14 make
   50  2022-10-17 17:53:14 nano Makefile
   51  2022-10-17 17:53:14 clear
   52  2022-10-17 17:53:14 ls
   53  2022-10-17 17:53:14 ls macro
   54  2022-10-17 17:53:14 nano macro/make.macro_gfortran-9_Linux
   55  2022-10-17 17:53:14 ls /usr/lib/gcc/x86_64-linux-gnu/
   56  2022-10-17 17:53:14 cp macro/make.macro_gfortran-9_Linux macro/make.macro_gfortran-pbow
   57  2022-10-17 17:53:14 cp macro/make.macro_gfortran-pbow make.macro
   58  2022-10-17 17:53:14 make
   59  2022-10-17 17:53:14 rm io_ezcdf.mod
   60  2022-10-17 17:53:14 rm -r mod
   61  2022-10-17 17:53:14 rm -r obj
   62  2022-10-17 17:53:14 ls
   63  2022-10-17 17:53:14 make
   64  2022-10-17 17:53:14 ls
   65  2022-10-17 17:53:14 nc-config --all
   66  2022-10-17 17:53:14 sudo apt-update nc-config
   67  2022-10-17 17:53:14 sudo apt update nc-config
   68  2022-10-17 17:53:14 sudo apt upgrade
   69  2022-10-17 17:53:14 nc-config --version
   70  2022-10-17 17:53:14 sudo apt-get update -y
   71  2022-10-17 17:53:14 sudo apt-get install -y netcdf-bin
   72  2022-10-17 17:53:14 make
   73  2022-10-17 17:53:14 nc-config -all
   74  2022-10-17 17:53:14 nc-config -version
   75  2022-10-17 17:53:14 nc-config --help
   76  2022-10-17 17:53:14 nc-config --version
   77  2022-10-17 17:53:14 nc-config --all
   78  2022-10-17 17:53:14 sudo apt-get install -y netcdf-bin
   79  2022-10-17 17:53:14 make
   80  2022-10-17 17:53:14 ls
   81  2022-10-17 17:53:14 ls obj
   82  2022-10-17 17:53:14 ls mod
   83  2022-10-17 17:53:14 nano make.macro
   84  2022-10-17 17:53:14 ls /usr/lib
   85  2022-10-17 17:53:14 nano make.macro
   86  2022-10-17 17:53:14 ls /usr/include
   87  2022-10-17 17:53:14 % cat netcdf.mod | gunzip | head -1
   88  2022-10-17 17:53:14 cat netcdf.mod | gunzip | head -1
   89  2022-10-17 17:53:14 sudo apt-get install libnetcdf-dev
   90  2022-10-17 17:53:14 which nc-config
   91  2022-10-17 17:53:14 nc-config
   92  2022-10-17 17:53:14 nc-config --all
   93  2022-10-17 17:53:14 ls /usr/bin/x86_64-linux-gnu-gcc
   94  2022-10-17 17:53:14 ls /usr/bin
   95  2022-10-17 17:53:14 nc-config --all
   96  2022-10-17 17:53:14 nc-config
   97  2022-10-17 17:53:14 nc-config --cc
   98  2022-10-17 17:53:14 nc-config --help
   99  2022-10-17 17:53:14 nf-config --help

CELTIC_SEA/make_tide.py
'''
this is an example to generate a ROMS tide file for TPXO8-atlas tidal product. Till this moment I couldn't found
a good python script to do so, so this script is created for this purpose. TPXO8-atlas is a high resolution global
model of ocean tides. it is available here:

http://volkov.oce.orst.edu/tides/tpxo8_atlas.html

download the 8 major constitutes with 1/30 deg resotuion. Other constitutes are also supported, but haven't been
tested. Only netCDF format is supported in this version, so make sure you download the right files.

The script 'make_remap_weights_file.py' is used to generate weight file between two grids, in this example the
CELTIC and TPXO8 grid files. TPXO8 grid file is kept in the same directory as data files:

/archive/u1/uaf/kate/tides/tpxo8/tpxo8_atlas_30c_v1.nc

Run this script before 'make_tide.py' to generate the weight files (remap_*.nc).

To use the remap function a new set of grid is created (CGrid_TPXO8). I followed the structure of GLORY CGrid
from the PYROMS packaage itself, so that it is consistent with PYROMS.

---------------------------------------------------------------------------------------------------------------------
tidal parameters are converted to elliptical parameters using ap2ep, origionally written in MATLAB and later
translated to python package PyFVCOM. See

CGrid_TPXO8/tidal_ellipse.py for more details.

-Chuning Wang, Dec 18 2016

'''

# File inspired by/modified from:
# /home/pete/PycharmProjects/pyroms_MI/examples/make_tide/make_tide.py
# Joe McGovern, Marine Institute, Rinville West, Rinville, Oranmore, Co. Galway 2021

import netCDF4 as nc
import numpy as np
import os
from scipy.interpolate import griddata

import pyroms
import pyroms_toolbox

# import examples.make_tide.CGrid_TPXO8.get_nc_CGrid_TPXO8
from examples.make_tide.CGrid_TPXO8 import get_nc_CGrid_TPXO8
from examples.make_tide.CGrid_TPXO8.tidal_ellipse import ap2ep
os.environ['PYROMS_GRIDID_FILE'] = '/home/pete/PycharmProjects/pyroms_MI/gridid_MI.txt'

# tidal constituents
consts1 =['Q1', 'O1', 'P1', 'K1', 'N2', 'M2', 'S2', 'K2']
consts2 = ['MF']
consts = consts1+consts2
# consts =['M2']
consts_num = len(consts)

# read ROMS grid
dstgrd = pyroms.grid.get_ROMS_grid('CELTIC')
lat = dstgrd.hgrid.lat_rho
lon = dstgrd.hgrid.lon_rho
msk = dstgrd.hgrid.mask_rho
eta, xi = lat.shape
# read weights file
wts_file_t = 'remap_weights_TPXO8_to_CELTIC_bilinear_t_to_rho.nc'
wts_file_u = 'remap_weights_TPXO8_to_CELTIC_bilinear_u_to_rho.nc'
wts_file_v = 'remap_weights_TPXO8_to_CELTIC_bilinear_v_to_rho.nc'

wts_file_t_lr = 'remap_weights_TPXO8atlas6_to_CELTIC_bilinear_t_to_rho.nc'
wts_file_u_lr = 'remap_weights_TPXO8atlas6_to_CELTIC_bilinear_u_to_rho.nc'
wts_file_v_lr = 'remap_weights_TPXO8atlas6_to_CELTIC_bilinear_v_to_rho.nc'

# read TPXO8 files
# pth_tpxo = '/archive/u1/uaf/kate/tides/tpxo8/'
# pth_tpxo = '/home/pete/PycharmProjects/pyroms_MI/Tideforce/TPX/TPX_08/'
# pth_tpxo9atlas = '/home/pete/PycharmProjects/pyroms_MI/Tideforce/TPX/TPX_09/ATLAS_06/'
# srcgrd = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo9_atlas_30_v2.nc ', xrange=(4400, 5100), yrange=(5200, 6500))
# srcgrd_lr = get_nc_CGrid_TPXO8(pth_tpxo9atlas+'grid_tpxo9.v2a.nc', name='TPXO8atlas6', xrange=(920, 1020),
#                                            yrange=(1030, 1300))
os.environ['PYROMS_GRIDID_FILE'] = '/home/pete/PycharmProjects/pyroms_MI/gridid_MI.txt'

pth_tpxo = '/home/pete/PycharmProjects/pyroms_MI/Tideforce/TPX/TPX_08/'
# srcgrd = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8atlas_30.nc', xrange=(4141, 4290), yrange=(121, 330))
# srcgrd_lr = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8_atlas6.nc', name='TPXO8atlas6',
#                                xrange=(829, 858), yrange=(25, 66))
# srcgrd = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8atlas_30.nc', xrange=(121, 330), yrange=(4141, 4290))
# srcgrd_lr = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8_atlas6.nc', name='TPXO8atlas6',
#                                xrange=(25, 66), yrange=(829, 858))
# srcgrd = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8atlas_30.nc', xrange=(4141, 4290), yrange=(121, 330))
# srcgrd_lr = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8_atlas6.nc', name='TPXO8atlas6',
#                                xrange=(829, 858), yrange=(25, 66))
srcgrd = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8atlas_30.nc', xrange=(4141, 4290), yrange=(5071, 5280))
srcgrd_lr = get_nc_CGrid_TPXO8(pth_tpxo+'grid_tpxo8_atlas6.nc', name='TPXO8atlas6',
                               xrange=(829, 858), yrange=(1009, 1056))
missing_value = srcgrd.missing_value

# grid sub-sample
xrange = srcgrd.xrange
yrange = srcgrd.yrange

xrange_lr = srcgrd_lr.xrange
yrange_lr = srcgrd_lr.yrange

# initiate variables
hamp = np.zeros((consts_num, eta, xi))*np.nan
hpha = np.zeros((consts_num, eta, xi))*np.nan
cmax = np.zeros((consts_num, eta, xi))*np.nan
cmin = np.zeros((consts_num, eta, xi))*np.nan
cang = np.zeros((consts_num, eta, xi))*np.nan
cpha = np.zeros((consts_num, eta, xi))*np.nan

k = -1
for cst in consts:
    k = k+1

    if cst in consts1:
        fid = nc.Dataset(pth_tpxo+'hf.'+cst.lower()+'_tpxo8_atlas_30c.nc', 'r')
        hRe = fid.variables['hRe'][yrange[0]:yrange[1]+1, xrange[0]:xrange[1]+1]
        hIm = fid.variables['hIm'][yrange[0]:yrange[1]+1, xrange[0]:xrange[1]+1]
    elif cst in consts2:
        fid = nc.Dataset(pth_tpxo+'hf.'+cst.lower()+'_tpxo8_atlas_6.nc', 'r')
        hRe = fid.variables['hRe'][yrange_lr[0]:yrange_lr[1]+1, xrange_lr[0]:xrange_lr[1]+1]
        hIm = fid.variables['hIm'][yrange_lr[0]:yrange_lr[1]+1, xrange_lr[0]:xrange_lr[1]+1]
    fid.close()

    if cst in consts1:
        fid = nc.Dataset(pth_tpxo+'uv.'+cst.lower()+'_tpxo8_atlas_30c.nc', 'r')
        uRe = fid.variables['uRe'][yrange[0]:yrange[1]+1, xrange[0]:xrange[1]+1]
        uIm = fid.variables['uIm'][yrange[0]:yrange[1]+1, xrange[0]:xrange[1]+1]
        vRe = fid.variables['vRe'][yrange[0]:yrange[1]+1, xrange[0]:xrange[1]+1]
        vIm = fid.variables['vIm'][yrange[0]:yrange[1]+1, xrange[0]:xrange[1]+1]
    elif cst in consts2:
        fid = nc.Dataset(pth_tpxo+'uv.'+cst.lower()+'_tpxo8_atlas_6.nc', 'r')
        uRe = fid.variables['uRe'][yrange_lr[0]:yrange_lr[1]+1, xrange_lr[0]:xrange_lr[1]+1]
        uIm = fid.variables['uIm'][yrange_lr[0]:yrange_lr[1]+1, xrange_lr[0]:xrange_lr[1]+1]
        vRe = fid.variables['vRe'][yrange_lr[0]:yrange_lr[1]+1, xrange_lr[0]:xrange_lr[1]+1]
        vIm = fid.variables['vIm'][yrange_lr[0]:yrange_lr[1]+1, xrange_lr[0]:xrange_lr[1]+1]
    fid.close()

    hRe = hRe*1.0
    hIm = hIm*1.0
    uRe = uRe*1.0
    uIm = uIm*1.0
    vRe = vRe*1.0
    vIm = vIm*1.0

    # unit convertion
    hRe = hRe/1000.  # mm to m
    hIm = hIm/1000.  # mm to m
    uRe = uRe/10000.  # cm2s-1 to m2s-1
    uIm = uIm/10000.  # cm2s-1 to m2s-1
    vRe = vRe/10000.  # cm2s-1 to m2s-1
    vIm = vIm/10000.  # cm2s-1 to m2s-1

    # convert vertical transport (uRe, vRe) to velocity
    if cst in consts1:
        zu = srcgrd.z_u
        zv = srcgrd.z_v
    elif cst in consts2:
        zu = srcgrd_lr.z_u
        zv = srcgrd_lr.z_v

    uRe = uRe/zu  # ms-1
    uIm = uIm/zu  # ms-1
    vRe = vRe/zv  # ms-1
    vIm = vIm/zv  # ms-1

    if cst in consts1:
        # mask invalid data
        hRe = np.ma.masked_where(~srcgrd.mask_t, hRe)
        hIm = np.ma.masked_where(~srcgrd.mask_t, hIm)
        uRe = np.ma.masked_where(~srcgrd.mask_u, uRe)
        uIm = np.ma.masked_where(~srcgrd.mask_u, uIm)
        vRe = np.ma.masked_where(~srcgrd.mask_v, vRe)
        vIm = np.ma.masked_where(~srcgrd.mask_v, vIm)

        # remap h, u, v
        hRe = pyroms.remapping.remap(hRe, wts_file_t, spval=missing_value)
        hIm = pyroms.remapping.remap(hIm, wts_file_t, spval=missing_value)
        uRe = pyroms.remapping.remap(uRe, wts_file_u, spval=missing_value)
        uIm = pyroms.remapping.remap(uIm, wts_file_u, spval=missing_value)
        vRe = pyroms.remapping.remap(vRe, wts_file_v, spval=missing_value)
        vIm = pyroms.remapping.remap(vIm, wts_file_v, spval=missing_value)

    elif cst in consts2:
        # mask invalid data
        hRe = np.ma.masked_where(~srcgrd_lr.mask_t, hRe)
        hIm = np.ma.masked_where(~srcgrd_lr.mask_t, hIm)
        uRe = np.ma.masked_where(~srcgrd_lr.mask_u, uRe)
        uIm = np.ma.masked_where(~srcgrd_lr.mask_u, uIm)
        vRe = np.ma.masked_where(~srcgrd_lr.mask_v, vRe)
        vIm = np.ma.masked_where(~srcgrd_lr.mask_v, vIm)

        # remap h, u, v
        hRe = pyroms.remapping.remap(hRe, wts_file_t_lr, spval=missing_value)
        hIm = pyroms.remapping.remap(hIm, wts_file_t_lr, spval=missing_value)
        uRe = pyroms.remapping.remap(uRe, wts_file_u_lr, spval=missing_value)
        uIm = pyroms.remapping.remap(uIm, wts_file_u_lr, spval=missing_value)
        vRe = pyroms.remapping.remap(vRe, wts_file_v_lr, spval=missing_value)
        vIm = pyroms.remapping.remap(vIm, wts_file_v_lr, spval=missing_value)

    hamp[k, :, :] = (hRe**2+hIm**2)**0.5  # mm
    hpha[k, :, :] = np.arctan(-hIm/hRe)/np.pi*180.  # deg
    uamp = (uRe**2+uIm**2)**0.5  # mm
    upha = np.arctan(-uIm/uRe)/np.pi*180.  # deg
    vamp = (vRe**2+vIm**2)**0.5  # mm
    vpha = np.arctan(-vIm/vRe)/np.pi*180.  # deg

    # convert ap to ep
    # cmax[k, :, :], ecc, cang[k, :, :], cpha[k, :, :], w = CGrid_TPXO8.tidal_ellipse.ap2ep(uamp, upha, vamp, vpha)
    cmax[k, :, :], ecc, cang[k, :, :], cpha[k, :, :], w = ap2ep(uamp, upha, vamp, vpha)
    cmin[k, :, :] = cmax[k, :, :]*ecc

# -------------------------------------------------------------------------
# mask land grid points
cmax[:, dstgrd.hgrid.mask_rho==0] = missing_value
cmin[:, dstgrd.hgrid.mask_rho==0] = missing_value
cang[:, dstgrd.hgrid.mask_rho==0] = missing_value
cpha[:, dstgrd.hgrid.mask_rho==0] = missing_value

# -------------------------------------------------------------------------
savedata = 1
if savedata == 1:
    # write tidal information to nc file
    # -------------------------------------------------------------------------
    # define tidal constituents names and periods
    tide_name = np.array([ list('Q1'), list('O1'), list('P1'), list('K1'),
                           list('N2'), list('M2'), list('S2'), list('K2'),
                           list('MF')])
    tide_period = np.array([26.8683567047119, 25.8193397521973, 24.0658893585205, 23.9344692230225,
                            12.6583499908447, 12.420599937439, 12, 11.9672346115112,
                            13.66079*24])

    # -------------------------------------------------------------------------
    # create nc file
    fh = nc.Dataset('Celtic_tides_otps.nc', 'w')
    fh.createDimension('namelen', 4)
    fh.createDimension('tide_period', consts_num)
    fh.createDimension('eta_rho', eta)
    fh.createDimension('xi_rho', xi)

    fh.history = 'Tides from TPXO8'
    import time
    fh.creation_date = time.strftime('%c')
    fh.Type = 'ROMS Tidal Forcing File'
    fh.Title = 'Forcing for' + dstgrd.name + 'domain'
    fh.grid_file = dstgrd.name
    fh.Source = 'OTPS'

    name_nc = fh.createVariable('tide_name', 'c', ('tide_period', 'namelen'))

    period_nc = fh.createVariable('tide_period', 'd', ('tide_period'), fill_value=-9999)
    period_nc.field = 'tide_period, scalar'
    period_nc.long_name = 'tidal angular period'
    period_nc.units = 'hours'

    lat_nc = fh.createVariable('lat_rho', 'd', ('eta_rho', 'xi_rho'))
    lat_nc.field = 'lat_rho, scalar'
    lat_nc.long_name = 'latitude of RHO-points'
    lat_nc.units = 'degree north'

    lon_nc = fh.createVariable('lon_rho', 'd', ('eta_rho', 'xi_rho'))
    lon_nc.field = 'lon_rho, scalar'
    lon_nc.long_name = 'longitude of RHO-points'
    lon_nc.units = 'degree east'

    msk_nc = fh.createVariable('mask_rho', 'd', ('eta_rho', 'xi_rho'))
    msk_nc.long_name = 'mask on RHO-points'
    msk_nc.option_0 = 'land'
    msk_nc.option_1 = 'water'

    Eamp_nc = fh.createVariable('tide_Eamp', 'd', ('tide_period', 'eta_rho', 'xi_rho'), fill_value=-9999)
    Eamp_nc.field = 'tide_Eamp, scalar'
    Eamp_nc.long_name = 'tidal elevation amplitude'
    Eamp_nc.units = 'meter'

    Ephase_nc = fh.createVariable('tide_Ephase', 'd', ('tide_period', 'eta_rho', 'xi_rho'), fill_value=-9999)
    Ephase_nc.field = 'tide_Ephase, scalar'
    Ephase_nc.long_name = 'tidal elevation phase angle'
    Ephase_nc.units = 'degrees, time of maximum elevation with respect chosen time orgin'

    Cmax_nc = fh.createVariable('tide_Cmax', 'd', ('tide_period', 'eta_rho', 'xi_rho'), fill_value=-9999)
    Cmax_nc.field = 'tide_Cmax, scalar'
    Cmax_nc.long_name = 'maximum tidal current, ellipse semi-major axis'
    Cmax_nc.units = 'meter second-1'

    Cmin_nc = fh.createVariable('tide_Cmin', 'd', ('tide_period', 'eta_rho', 'xi_rho'), fill_value=-9999)
    Cmin_nc.field = 'tide_Cmin, scalar'
    Cmin_nc.long_name = 'minimum tidal current, ellipse semi-minor axis'
    Cmin_nc.units = 'meter second-1'

    Cangle_nc = fh.createVariable('tide_Cangle', 'd', ('tide_period', 'eta_rho', 'xi_rho'), fill_value=-9999)
    Cangle_nc.field = 'tide_Cangle, scalar'
    Cangle_nc.long_name = 'tidal current inclination angle'
    Cangle_nc.units = 'degrees between semi-major axis and East'

    Cphase_nc = fh.createVariable('tide_Cphase', 'd', ('tide_period', 'eta_rho', 'xi_rho'), fill_value=-9999)
    Cphase_nc.field = 'tide_Cphase, scalar'
    Cphase_nc.long_name = 'tidal current phase angle'
    Cphase_nc.units = 'degrees, time of maximum velocity'

    # -------------------------------------------------------------------------
    # write data
    name_nc[:, 0:2] = tide_name
    period_nc[:] = tide_period

    lat_nc[:, :] = lat
    lon_nc[:, :] = lon
    msk_nc[:, :] = msk
    Eamp_nc[:, :, :] = hamp
    Ephase_nc[:, :, :] = hpha
    Cmax_nc[:, :, :] = cmax
    Cmin_nc[:, :, :] = cmin
    Cangle_nc[:, :, :] = cang
    Cphase_nc[:, :, :] = cpha

    fh.close()
